vyper/context/definitions/builtin_functions.py

```diff
@@ -0,0 +1,623 @@
+
+
+class MethodID(BuiltinFunctionDefinition):
+
+    _id = "method_id"
+
+    def evaluate(self, node: vy_ast.Call) -> vy_ast.Bytes:
+        self._validate_call(node)
+
+        sig = get_literal_or_raise(node.args[0]).value
+        length = get_type_from_annotation(node.args[1]).length
+
+        method_id = fourbytes_to_int(keccak256(sig.encode())[:4])
+        value = method_id.to_bytes(length, "big")
+
+        return vy_ast.Bytes.from_node(node, value=value)
+
+    def _validate_call(self, node) -> None:
+        validate_call_args(node, 2)
+        sig = get_literal_or_raise(node.args[0]).value
+        if not isinstance(node.args[0], vy_ast.Str):
+            raise InvalidType("method id must be given as a literal string", node.args[0])
+        if " " in sig:
+            raise TypeMismatch('Invalid function signature no spaces allowed.')
+        return_type = get_type_from_annotation(node.args[1])
+        if not is_subtype(return_type, BytesBase) or return_type.length not in (4, 32):
+            raise InvalidType("return type must be bytes32 or bytes[4]", node.args[1])
+
+    def fetch_call_return(self, node: vy_ast.Call) -> None:
+        self._validate_call(node)
+        # this method should always be foldable - any invalid user input
+        # should raise from one of the previous checks
+        raise CompilerPanic("method_id should have been folded")
+
+
+class Extract32(BuiltinFunctionDefinition):
+
+    _id = "extract32"
+
+    def fetch_call_return(self, node: vy_ast.Call) -> Reference:
+        validate_call_args(node, 2, ["type"])
+        target, length = (get_definition_from_node(i).type for i in node.args[:2])
+
+        compare_types(target, namespace['bytes'], node.args[0])
+        compare_types(length, namespace['int128'], node.args[1])
+
+        if node.keywords:
+            return_type = get_type_from_annotation(node.keywords[0].value)
+            expected = get_builtin_type({"address", "bytes32", "int128", "uint256"})
+            compare_types(return_type, expected, node.keywords[0])
+
+        else:
+            return_type = get_builtin_type("bytes32")
+
+        return Reference.from_type(return_type, "return value")
+
+
+class RawLog(BuiltinFunctionDefinition):
+
+    _id = "raw_log"
+
+    def fetch_call_return(self, node: vy_ast.Call) -> None:
+        validate_call_args(node, 2)
+        if not isinstance(node.args[0], vy_ast.List) or len(node.args[0].elts) > 4:
+            raise InvalidType(
+                "Expecting a list of 0-4 topics as first argument", node.args[0]
+            )
+        if node.args[0].elts:
+            log_type = get_definition_from_node(node.args[0]).type
+            compare_types(log_type[0], namespace['bytes32'], node.args[0])
+        compare_types(
+            get_definition_from_node(node.args[1]).type,
+            get_builtin_type({'bytes', 'bytes32'}),
+            node.args[1]
+        )
+        return None
+
+
+class Convert(BuiltinFunctionDefinition):
```

Comment:
I wonder if this definition deserves it's own file as it's likely to be pretty large

Unless the conversion logic is handled in other nodes via a common API

Reply:
I don't like the location of this entire module.  For most of the project I like the approach of one pass, one subpackage, but for functions I feel it would be better organized if all their logic start-to-finish is in the same place.  That way it's easy to find, you can grok how a function is type-checked, folded, and converted to LLL without having to jump all over the codebase.

The obvious choice for where this lives is `vyper/functions`, but that already exists. So I left it here temporarily, I figure once I get to removing redundant code I can move it.

If you have any other ideas I'm open to them.